app.py

```diff
@@ -70,6 +71,51 @@
 
 app = Flask(__name__, static_url_path='')
 
+def hash_user_or_session (string):
+    hash = hashlib.md5 (string.encode ('utf-8')).hexdigest ()
+    return int (hash, 16)
+
+def redirect_ab (request, session):
+    # If the user is logged in, we use their username as identifier, otherwise we use the session id
+    user_identifier = current_user(request) ['username'] or str (session_id ())
+
+    # This will send 50% of the requests into redirect.
+    redirect_proportion = 50
+    redirect_flag = (hash_user_or_session (user_identifier) % 100) < redirect_proportion
+    return redirect_flag
+
+# If present, REDIRECT_AB_TEST should be the name of the target environment (that is, environment B).
+if os.getenv ('REDIRECT_AB_TEST') and os.getenv ('REDIRECT_AB_TEST') != os.getenv ('HEROKU_APP_NAME'):
+    @app.before_request
+    def before_request():
+        # If it is an auth route, we do not reverse proxy it to the REDIRECT_AB_TEST environment, with the exception of /auth/texts
+        # We want to keep all cookie setting in the main environment, not the test one.
+        if (re.match ('.*/auth/.*', request.url) and not re.match ('.*/auth/texts', request.url)):
+            # Dummy value. We do this in a separate block so that the block below doesn't depend on a very long conditional.
+            0
+        # If we enter this block, we will reverse proxy the request to the REDIRECT_AB_TEST environment.
+        elif (redirect_ab (request, session)):
+            url = request.url.replace (os.getenv ('HEROKU_APP_NAME'), os.getenv ('REDIRECT_AB_TEST'))
+
+            request_headers = {}
+            for header in request.headers:
+                if (header [0].lower () in ['host']):
+                    continue
+                request_headers [header [0]] = header [1]
+
+            # Send the session_id to the test environment for logging purposes
+            request_headers ['x-session_id'] = session_id ()
```

Comment:
I like this! But please do document the private protocol that is now being used between the two Hedy instances and how cookies are handled somewhere.

Bottom line:

- Cookies are only set on the primary Hedy instance (both for sessions and login)
- Session IDs are forwarded in the reverse proxy protocol, and taken as gospel in the test instance
- Login cookies are forwarded and looked up the database by both instances
- Session cookies are still uselessly forwarded though, right? We should probably disable session on the Hedy test instance? Or something? Or have it not try to start sessions on its own, for no purpose?